packages/devtools_app/lib/src/timeline/timeline_controller.dart

```diff
@@ -74,123 +59,77 @@ class TimelineController {
+      : fullTimeline.data;
 
-    if (debugTimeline && frame != null) {
-      final buf = StringBuffer();
-      buf.writeln('UI timeline event for frame ${frame.id}:');
-      frame.uiEventFlow.format(buf, '  ');
-      buf.writeln('\nUI trace for frame ${frame.id}');
-      frame.uiEventFlow.writeTraceToBuffer(buf);
-      buf.writeln('\nGPU timeline event frame ${frame.id}:');
-      frame.gpuEventFlow.format(buf, '  ');
-      buf.writeln('\nGPU trace for frame ${frame.id}');
-      frame.gpuEventFlow.writeTraceToBuffer(buf);
-      log(buf.toString());
-    }
-  }
+  CpuProfileData get cpuProfileData =>
+      timelineMode == TimelineMode.frameBased || offlineMode
+          ? frameBasedTimeline.data?.cpuProfileData
+          : fullTimeline.data?.cpuProfileData;
 
   void selectTimelineEvent(TimelineEvent event) {
-    if (event == null || timelineData.selectedEvent == event) {
-      return;
+    if (event == null) return;
+    if (timelineMode == TimelineMode.frameBased) {
+      if (frameBasedTimeline.data.selectedEvent == event) return;
+      frameBasedTimeline.data.selectedEvent = event;
+    } else {
+      if (fullTimeline.data.selectedEvent == event) return;
+      fullTimeline.data.selectedEvent = event;
     }
-    timelineData.selectedEvent = event;
     _selectedTimelineEventController.add(event);
   }
 
-  void addFrame(TimelineFrame frame) {
-    timelineData.frames.add(frame);
-    frameAddedController.add(frame);
-  }
-
   Future<void> getCpuProfileForSelectedEvent() async {
-    if (!timelineData.selectedEvent.isUiEvent) return;
-
-    assert(timelineData.selectedEvent.frameId == timelineData.selectedFrame.id);
+    final selectedEvent = timelineData.selectedEvent;
+    if (!selectedEvent.isUiEvent) return;
 
-    timelineData.selectedFrame.cpuProfileData ??=
-        await _cpuProfilerService.getCpuProfile(
-      startMicros:
-          timelineData.selectedFrame.uiEventFlow.time.start.inMicroseconds,
-      extentMicros:
-          timelineData.selectedFrame.uiEventFlow.time.duration.inMicroseconds,
+    final cpuProfileData = await _cpuProfilerService.getCpuProfile(
+      startMicros: selectedEvent.time.start.inMicroseconds,
+      extentMicros: selectedEvent.time.duration.inMicroseconds,
     );
 
-    timelineData.cpuProfileData = CpuProfileData.subProfile(
-      timelineData.selectedFrame.cpuProfileData,
-      timelineData.selectedEvent.time,
-    );
-    _cpuProfileTransformer.processData(timelineData.cpuProfileData);
-  }
-
-  void recordTrace(Map<String, dynamic> trace) {
-    timelineData.traceEvents.add(trace);
-  }
-
-  void recordTraceForTimelineEvent(TimelineEvent event) {
-    recordTrace(event.beginTraceEventJson);
-    event.children.forEach(recordTraceForTimelineEvent);
-    if (event.endTraceEventJson != null) {
-      recordTrace(event.endTraceEventJson);
+    if (timelineMode == TimelineMode.frameBased) {
+      frameBasedTimeline.data.cpuProfileData = cpuProfileData;
```

Comment:
have a getter for the current timeline and use that everywhere you need this sort of logic instead of using an iff statement each case you need to condition based on the current mode.